### PR TITLE
Redesign Xauth handling

### PR DESCRIPTION
**Pull request by Vogtinator**
This commit moves Xauthority handling over to libXau.
Advantage is that this allows use of FamilyWild, is faster, more reliable
and easier to read. However, we lose the ability to merge the new cookie into
an existing Xauthority file, so support for using a non-temporary file is
dropped. Even if merging was implemented manually, use of FamilyWild would
"infect" such a file and break it for DMs which don't write it.

### CODE REVIEW

File: data/man/sddm.conf.rst.in
Comment (plfiorini) [inline data/man/sddm.conf.rst.in:126]: I think that we should add a paragraph to inform the users that XauthPath and UserAuthFile are dead now and why.

File: src/helper/UserSession.cpp
Comment (plfiorini) [inline src/helper/UserSession.cpp]: // comments

**Comment by tpgxyz** [comment]
I've applied this PR to https://github.com/OpenMandrivaAssociation/sddm/commit/5384de223c4885267e48d09de0a36b25fc23dc87

I'm running sddm on OpenMandriva with this PR, found no issues.

File: src/helper/UserSession.cpp
Comment (davidedmundson) [inline src/helper/UserSession.cpp]: This is somewhat weird.

setupChildProcess is after we've forked, we have the same memory contents, but it's not shared. It's COW.

Does it work as intended and actually clean up?

I suspect it would seem to work because:

the m_xauthFile of the parent that's actually scoped to the process stays effectively uninitialised

the child sets up m_xauthFile but in it's cloned memory, we would then effectively leak the temporary file as we don't call the destructor when we exec()

File: src/common/XauthUtils.cpp
Comment (plfiorini) [inline src/common/XauthUtils.cpp]: please add the author with SPDX tags

```
/****************************************************************************
 * SPDX-FileCopyrightText: 2020 NAME <email>
 *
 * SPDX-License-Identifier: GPL-2.0-or-later
 ***************************************************************************/
```

File: src/daemon/XorgDisplayServer.cpp
Comment (plfiorini) [inline src/daemon/XorgDisplayServer.cpp]: is returning const QByteArray & necessary, why not simply QByteArray

File: src/helper/UserSession.cpp
Comment (Vogtinator) [reply src/helper/UserSession.cpp]: Good catch - I totally forgot to mention that.

It leaks the temporary file currently, so relies on `/run` getting cleaned up by something else.

Not sure how to improve that though, the deletion would need to happen in the user context as well to avoid any arbitrary deletion vulnerabilities.

There's also https://github.com/sddm/sddm/issues/1083 which needs a way to clean up stuff after a user session ends.

File: src/daemon/XorgDisplayServer.cpp
Comment (Vogtinator) [reply src/daemon/XorgDisplayServer.cpp]: it was the same for QString before, so I just kept it.

**Comment by Vogtinator** [comment]
> plfiorini approved these changes 1 minute ago

Note that there are currently open topics, like the authority file leak and the manpage update.

File: src/helper/UserSession.cpp
Comment (Vogtinator) [reply src/helper/UserSession.cpp]: What about redesigning `sddm-helper` away from `setupChildProcess`? It could just do the setup in the main process itself (like `setuid`) and then use `QProcess` as usual. This means that `sddm-helper` would effectively run as the target user at that point and proper cleanup can be performed.

It would not help with #1083 though as that's managed by `sddm`, not `sddm-helper` IIRC.

Comment (Vogtinator) [reply src/helper/UserSession.cpp]: This is technically possible, but currently `sddm-helper` performs utmp and PAM session handling after the user process quites, so that would have to be moved into the daemon first...

I tested how this behaves on FreeBSD and there the session setup is  done as part of `Xsession`: https://github.com/freebsd/freebsd-ports-kde/blob/d70b3c9023dfe8414ea91c31881209d16c15b229/x11/sddm/files/patch-data_scripts_Xsession#L18
This also includes setting up `XDG_RUNTIME_DIR`, which means that it's not available inside `sddm-helper` at all.
While using the fallback directory there is technically fine, it definitely needs manual cleanup then.

File: data/man/sddm.conf.rst.in
Comment (Vogtinator) [reply data/man/sddm.conf.rst.in:126]: Now in the `sddm` page

**Comment by Vogtinator** [comment]
I changed the design somewhat. Instead of putting the temporary `xauth_XXXXX` file into the user's runtime dir, it now ends up in sddm's `RUNTIME_DIR` and gets its ownership changed to the target user. This means that cleanup by `sddm-helper` is possible now and it should also work on FreeBSD.

**Comment by mcsurly** [comment]
Placing file xauth_XXXXX has bad side-effects. The user cannot write to the containing directory, so xauth (the program) cannot create the lockfile RUNTIME_DIR/xauth_XXXXX-c
(it appends -c to the original filename and tries to create the lock file)
This makes all executions of xauth take 20 seconds to time out (and set RC=1). This is most painful whenever you ssh with X forwarding enabled. 
See downstream bug https://bugzilla.redhat.com/show_bug.cgi?id=1922772

> I changed the design somewhat. Instead of putting the temporary `xauth_XXXXX` file into the user's runtime dir, it now ends up in sddm's `RUNTIME_DIR` and gets its ownership changed to the target user. This means that cleanup by `sddm-helper` is possible now and it should also work on FreeBSD.

**Comment by tpgxyz** [comment]
@tvolin make sure you have set proper permissions on `/run/sddm`
see here https://github.com/OpenMandrivaAssociation/sddm/commit/3cf56782a353f6187a7007862d51961bf706df77

**Comment by Vogtinator** [comment]
> Placing file xauth_XXXXX has bad side-effects. The user cannot write to the containing directory, so xauth (the program) cannot create the lockfile RUNTIME_DIR/xauth_XXXXX-c
> (it appends -c to the original filename and tries to create the lock file)
> This makes all executions of xauth take 20 seconds to time out (and set RC=1). This is most painful whenever you ssh with X forwarding enabled.
> See downstream bug https://bugzilla.redhat.com/show_bug.cgi?id=1922772

Great, back to the drawing board. Apparently this approach doesn't work either then.

**Comment by Vogtinator** [comment]
> @tvolin make sure you have set proper permissions on `/run/sddm`
> see here [OpenMandrivaAssociation/sddm@3cf5678](https://github.com/OpenMandrivaAssociation/sddm/commit/3cf56782a353f6187a7007862d51961bf706df77)

O.o

You realize that way you're giving EVERY USER on the system FULL ACCESS to your sessions and the login screen, which means they can just run xev and grab your passwords?

**Comment by tpgxyz** [comment]
> You realize that way you're giving EVERY USER on the system FULL ACCESS to your sessions and the login screen, which means they can just run xev and grab your passwords?

oops, thanks for spotting this ⭕

**Comment by Vogtinator** [comment]
Actually, I missed that the directory has the sticky bit set - in that case it should actually be fine (but no guarantee).

There are other issues with this approach though, for instance that there's still no explicit cleanup in case of leftover files (sddm crash, power outage, etc.).

**Comment by tvolin** [comment]
Permissions 1733 seem to do the job. Non-root users can't even see the name of each other's xauth files. And with 0600 as the file permission (which it is), they cannot read each other's files either.
Cleanup of /var/run on Fedora is a non-issue since it's on tmpfs. Not sure about other distros.
Bob Richmond proposed this solution for Fedora. Since Mandriva did something similar I'll suggest we fix this downstream.

**Comment by Vogtinator** [comment]
> Permissions 1733 seem to do the job. Non-root users can't even see the name of each other's xauth files. And with 0600 as the file permission (which it is), they cannot read each other's files either.
> Cleanup of /var/run on Fedora is a non-issue since it's on tmpfs. Not sure about other distros.

I'm wondering mostly about FreeBSD actually.

> Bob Richmond proposed this solution for Fedora. Since Mandriva did something similar I'll suggest we fix this downstream.

**Comment by rdieter** [comment]
ping, still anxiously waiting for this to get approvied/merged, so one less thing to carry downstream patch for.

**Comment by Vogtinator** [comment]
What about just putting everything into `/tmp` directly? That's where it puts the sockets already and `/tmp` already has the proper permissions to store xauth files inside. As it's just a single file per session, it wouldn't cause any pollution either. It might need an entry in `tmpfiles.d` though.

**Comment by Vogtinator** [comment]
> What about just putting everything into `/tmp` directly? That's where it puts the sockets already and `/tmp` already has the proper permissions to store xauth files inside. As it's just a single file per session, it wouldn't cause any pollution either. It might need an entry in `tmpfiles.d` though.

I pushed a new version doing just that, it seems to work in the (limited) tests I did.

**Comment by Pointedstick** [comment]
Not sure I am qualified to review this, sorry. I will leave it to others. :)

**Review by Conan-Kudo — APPROVED** [review]
We've been running with this patch in Fedora for a year now, and it's been fine for us.

**Comment by Conan-Kudo** [comment]
@Vogtinator Can you rebase this on the current mainline to make it mergeable?

**Comment by Vogtinator** [comment]
> @Vogtinator Can you rebase this on the current mainline to make it mergeable?

Well, I did yesterday but just shortly afterwards major changes got merged...

I don't currently have time to get familiar with them and redo this PR.

**Comment by tpgxyz** [comment]
@Vogtinator can you please rebase it on top current head ?

**Comment by Conan-Kudo** [comment]
@Vogtinator @davidedmundson Can someone please look at rebasing this on top of current mainline? We need this and I couldn't figure out how to do it myself. 😢

**Comment by Vogtinator** [comment]
Tested to work as expected with `x11`, `x11-user` and `wayland`. For `x11`, `XAUTHORITY` is set to `/tmp/xauth_rAnDoM`, for `x11-user` to `/run/user/1337/xauth_rAnDoM`.

File: src/common/XAuth.cpp
Comment (davidedmundson) [inline src/common/XAuth.cpp]: I would suggest using QScopeGuard rather than having to fClose in all the paths

**Review by davidedmundson — APPROVED** [review]
Looks good. Thanks

File: src/common/XAuth.cpp
Comment (Vogtinator) [reply src/common/XAuth.cpp]: Done

**Pull request merged**
